guava/src/com/google/common/io/LineReader.java

```diff
@@ -84,4 +94,43 @@ public String readLine() throws IOException {
     }
     return lines.poll();
   }
+
+    /**
+     * Returns a {@code Stream}, the elements of which are lines read from
+     * this {@code LineReader}.
+     *
+     * @return a {@code Stream<String>} providing the lines of text
+     *         described by this {@code LineReader}
+     */
+    public Stream<String> lines()
+    {
+        Iterator<String> iterator = new Iterator<String>() {
```

Comment:
Perhaps using the [AbstractIterator](https://github.com/google/guava/blob/master/guava/src/com/google/common/collect/AbstractIterator.java) would make the logic quite a bit simpler